streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -347,8 +347,9 @@ private boolean setState(final State newState) {
             } else if (state == State.REBALANCING && newState == State.REBALANCING) {
                 // when the state is already in REBALANCING, it should not transit to REBALANCING again
                 return false;
-            } else if (state == State.ERROR && (newState == State.PENDING_ERROR || newState == State.ERROR)) {
-                // when the state is already in ERROR, its transition to PENDING_ERROR or ERROR (due to consecutive close calls)
+            } else if (state == State.ERROR && (newState == State.PENDING_ERROR || newState == State.ERROR || newState == State.PENDING_SHUTDOWN)) {
```

Comment:
We need transitions of all shutdown states (`PENDING_SHUTDOWN`, `NOT_RUNNING`, `ERROR`, `PENDING_ERROR`) to `PENDING_SHUTDOWN` to be non-fatal (rejecting transition but not throwing), so that we can attempt to transition. The transition from `ERROR` to `PENDING_SHUTDOWN` was the only fatal combination. This wasn't triggered in the original code but would be triggered by the fix.

Reply:
> We would still miss PENDING_ERROR -> PENDING_SHUTDOWN below

This is covered by
`} else if (state == State.PENDING_ERROR && newState != State.ERROR) {`

> as well as NOT_RUNNING -> PENDING_SHUTDOWN` above?

This is covered by
`} else if (state == State.NOT_RUNNING && (newState == State.PENDING_SHUTDOWN || newState == State.NOT_RUNNING)) {`